simd/src/Kokkos_Simd_Common.hpp

```diff
@@ -0,0 +1,224 @@
+#ifndef KOKKOS_SIMD_COMMON_HPP
+#define KOKKOS_SIMD_COMMON_HPP
+
+#include <cmath>
+#include <cstring>
+
+#include <Kokkos_Macros.hpp>
+
+#include <desul/atomics/Common.hpp>
+
+namespace Kokkos {
+namespace Experimental {
+
+template <class To, class From>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION constexpr To bit_cast(
+    From const& src) {
+  To dst;
+  std::memcpy(&dst, &src, sizeof(To));
+  return dst;
+}
+
+template <class T, class Abi>
+class simd;
+
+template <class T, class Abi>
+class simd_mask;
+
+class element_aligned_tag {};
+
+template <class Mask, class Value>
+class const_where_expression {
+ protected:
+  Value& m_value;
+  Mask const& m_mask;
+
+ public:
+  const_where_expression(Mask const& mask_arg, Value const& value_arg)
+      : m_value(const_cast<Value&>(value_arg)), m_mask(mask_arg) {}
+};
+
+template <class Mask, class Value>
+class where_expression : public const_where_expression<Mask, Value> {
+  using base_type = const_where_expression<Mask, Value>;
+
+ public:
+  where_expression(Mask const& mask_arg, Value& value_arg)
+      : base_type(mask_arg, value_arg) {}
+};
+
+template <class T, class Abi>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION where_expression<simd_mask<T, Abi>,
+                                                           simd<T, Abi>>
+where(desul::Impl::dont_deduce_this_parameter_t<simd_mask<T, Abi>> const& mask,
+      simd<T, Abi>& value) {
+  return where_expression(mask, value);
+}
+
+template <class T, class Abi>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION const_where_expression<
```

Comment:
Why "force inline"?  (Here and everywhere below)

Reply:
In the P3A library that I copied this from, I found that this made a significant performance difference over just "inline" for other methods in the code (for example functor `operator()`), so I've gotten into the habit of marking anything that should ideally disappear as a compile-time abstraction with force inline. This gives me peace of mind that calls to `simd` types will decay into just intrinsic instruction calls.